### PR TITLE
[LI-HOTFIX] CI: Split test builds to multiple jobs

### PR DESCRIPTION
**Pull request by lmr3796**
LI_DESCRIPTION =
Previously we've experienced some patches passed unit test build but failed the main release build,
so we want to run integration tests on PR too.  However, running all tests at once
makes the PR feedback time too long, so we split them to subsets (by the prefix of test name)
and put them to multiple jobs for faster feedback.

The retry mechanics is unfortunately only available in 2.7+, because it
requires junit >= 5.7 (with experiment at least 4.13 doesn't work).

TICKET = N/A
EXIT_CRITERIA = N/A

### CODE REVIEW

**Review by efeg — APPROVED** [review]

File: .github/workflows/build-and-test-on-push-or-pr.yml [hunk 1]
```diff
@@ -42,4 +44,5 @@ jobs:
           java-version: 1.8
       - name: Build with Gradle and run all unit tests
         # exclude the streams test and connect test
-        run: ./gradlew cleanTest rat checkstyleMain checkstyleTest spotbugsMain spotbugsTest :clients:unitTest :core:unitTest --no-daemon -PxmlFindBugsReport=true -PtestLoggingEvents=started,passed,skipped,failed
+        # Set maxTestRetries & maxTestRetryFailures for flaky tests.  This is the setup in upstream Jenkinsfile
+        run: ./gradlew -PmaxTestRetries=1 -PmaxTestRetryFailures=5 cleanTest rat checkstyleMain checkstyleTest spotbugsMain spotbugsTest :clients:test :core:test --no-daemon -PxmlFindBugsReport=true -PtestLoggingEvents=started,passed,skipped,failed
```
Comment (gitlw) [inline .github/workflows/build-and-test-on-push-or-pr.yml]: As discussed offline, we'll explore splitting the tests into multiple tasks in Github as a way to reduce the turnaround time for PR updates. I think if the time can be reduced to be under 30min, it'd be acceptable to run all the tests for PR updates.

File: .github/workflows/unit-test.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License
+
+name: unitTest
+
+on:
+  pull_request:
```
Comment (gitlw) [inline .github/workflows/unit-test.yml:21]: Should we still keep the unit tests for PR updates?

Comment (lmr3796) [reply .github/workflows/unit-test.yml:21]: Yes we still keep it --- see the executed checks below.

The empty `pull_request` object actually means "no additional setup to pull_request events, just listen to it".  Some would set options under `pull_request` object to tweak the details.

File: .github/workflows/build-and-test-on-push-or-pr.yml (hunk 1 above)
Comment (lmr3796) [reply .github/workflows/build-and-test-on-push-or-pr.yml]: In the setup we split them to 5 actions (and they're actually executed).   Unfortunately it's the single `core:integrationTest` taking most of the time.

File: .github/workflows/int-test.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+          ref: ${{ github.event.pull_request.head.sha }}
+          # bring in all history because the gradle versions plugin needs to "walk back" to the closest ancestor tag
+          fetch-depth: 0
+      - name: Set up JDK 8
+        uses: actions/setup-java@v2
+        with:
+          java-version: '8'
+          distribution: 'zulu'
+      - name: Run tests
+        run: ./gradlew -PmaxTestRetries=3 cleanTest core:integrationTest --no-daemon -PtestLoggingEvents=started,passed,skipped,failed
```
Comment (gitlw) [inline .github/workflows/int-test.yml]: Do you want to break down the set of IntegrationTests further, e.g. one group of tests that start with a letter in the range a-m, and another group that start with a letter in the range n-z?
After parallelizing, if we can bring down the test down below 30min, it'll make sense to include it as part of the code review flow.

File: .github/workflows/build-and-test-on-push-or-pr.yml (hunk 1 above)
Comment (lmr3796) [reply .github/workflows/build-and-test-on-push-or-pr.yml]: The test is always executed, but maybe we don't have to wait on the 1hr long `core:integrationTest` to finish on PR.

File: .github/workflows/unit-test.yml (hunk 2 above)
Comment (gitlw) [reply .github/workflows/unit-test.yml:21]: Ah, thanks for the explanation.

File: .github/workflows/int-test.yml (hunk 3 above)
Comment (lmr3796) [reply .github/workflows/int-test.yml]: I've broken down the tests into multiple subsets (partitioned by the first letter of class names).  This should make the turn around time faster.  However, retry plugin is not available until 2.7, which brings junit 5 required by the retry plugin.

**Review by gitlw — APPROVED** [review]
LGTM, thanks for this change!

**Pull request merged**